src/NCMenuBar.cc

```diff
@@ -0,0 +1,217 @@
+/*
+  Copyright (C) 2020 SUSE LLC
+  This library is free software; you can redistribute it and/or modify
+  it under the terms of the GNU Lesser General Public License as
+  published by the Free Software Foundation; either version 2.1 of the
+  License, or (at your option) version 3.0 of the License. This library
+  is distributed in the hope that it will be useful, but WITHOUT ANY
+  WARRANTY; without even the implied warranty of MERCHANTABILITY or
+  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+  License for more details. You should have received a copy of the GNU
+  Lesser General Public License along with this library; if not, write
+  to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
+  Floor, Boston, MA 02110-1301 USA
+*/
+
+
+/*-/
+
+   File:       NCMenuBar.cc
+
+   Author:     Stefan Hundhammer <shundhammer@suse.de>
+
+/-*/
+
+
+#include <algorithm>    // std::max()
+
+#define	 YUILogComponent "ncurses"
+#include <yui/YUILog.h>
+#include "NCMenuBar.h"
+#include "NCPulldownMenu.h"
+#include "YNCursesUI.h"
+
+
+// Margin to the left of the first toplevel item
+#define LEFT_MARGIN     0
+
+// Spacing between toplevel items
+#define SPACING         0
+
+using std::string;
+
+
+NCMenuBar::NCMenuBar( YWidget * parent )
+    : YMenuBar( parent )
+    , NCWidget( parent )
+{
+    // yuiDebug() << endl;
+    wstate = NC::WSdumb;
+}
+
+
+NCMenuBar::~NCMenuBar()
+{
+    // yuiDebug() << endl;
+
+    clear();
+}
+
+
+void
+NCMenuBar::clear()
+{
+    for ( unsigned i=0; i < _menus.size(); ++i )
+        delete _menus[i];
+
+    _menus.clear();
+    defsze = wsze( 1, 10 );
+}
+
+
+void
+NCMenuBar::rebuildMenuTree()
+{
+    clear();
+    int width = 0;
+
+    for ( YItemIterator it = itemsBegin(); it != itemsEnd(); ++it )
+    {
+	YMenuItem * item = dynamic_cast<YMenuItem *>( *it );
+        YUI_CHECK_PTR( item );
+
+        if ( ! item->isMenu() )
+            YUI_THROW( YUIException( "NCMenuBar: Only menus allowed on toplevel." ) );
+
+        yuiDebug() << "Adding " << item->label() << endl;
+        NCPulldownMenu * menu = new NCPulldownMenu( this, item->label() );
+        item->setUiItem( menu );
+        _menus.push_back( menu );
+
+        if ( width > 0 )
+            width += SPACING;
+
+        width += menu->preferredWidth();
+    }
+
+    width += LEFT_MARGIN;
+    defsze = wsze( 1, width );
+
+    // yuiDebug() << "Finish" << endl;
+}
+
+
+NCursesEvent
+NCMenuBar::postMenu()
+{
+    NCursesEvent ret;
+    // TO DO
+
+    return ret;
+}
+
+
+void
+NCMenuBar::setItemEnabled( YMenuItem * item, bool enabled )
+{
+    // TO DO
+}
+
+
+void
+NCMenuBar::activateItem( YMenuItem * item )
+{
+    // TO DO
+}
+
+
+NCursesEvent
+NCMenuBar::wHandleInput( wint_t key )
```

Comment:
According to my manual testing, this method is not call when the focus is over a `NCPulldownMenu`. But `NCPulldownMenu::vHanleInput` is properly called (after adding it). Any clue about this?

I guess we need to handle the input from the `NCMenuBar`, otherwise it would be difficult to implement the movement between the most top menu options.

Reply:
I don't know; I didn't get that far yet. I was glad to see the layout and visual appearance if those pulldown menus working before my vacation. It is very well possible that most event handling will have to be done in the pulldown menu event handlers, or that they have to delegate some events up to their parent menu bar.